### ISSUE TITLE
rule [no-invalid-this] and [new function] don't correctly work

### ISSUE THREAD

**Issue by boggddan** [eslint/eslint#8863]
Hello,
Rule **no-invalid-this** and **new function** inside in function or class, Eslint error message.

**new function** - is the function constructor, and have containing **this**

But eslint 

Samples: 
```javascript
/eslint no-invalid-this: "error"/
/eslint-env es6/
'use strict';

function test() {
  const aObj = new function( text ) {
    this.test = text; // this eslint error "no-ivalid-this"
  }( 'test' );
  return aObj;
}
```

But, If create function-constructor, when not error.   

```javascript
/eslint-env es6/
'use strict';

function test() {
  function New( text ) {
    this.test = text; // this eslint error "no-ivalid-this"
  }
  const aObj = new New( 'test' );
  return aObj;
}
```

But need create function when used once, there using **new function**.

**Comment by eslintbot** [comment]
Hi @boggddan, thanks for the issue. It looks like there's not enough information for us to know how to help you.

If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

Requesting a rule change? Please see [Proposing a Rule Change](http://eslint.org/docs/developer-guide/contributing/rule-changes) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by kaicataldo** [comment]
Thanks for the issue. I'm a bit confused by your second example code block:

```js
/* eslint-env es6 */
'use strict';

function test() {
  function New( text ) {
    this.test = text; // this eslint error "no-ivalid-this"
  }
  const aObj = new New( 'test' );
  return aObj;
}
```

`this.test = text; // this eslint error "no-invalid-this"` does not error for me with the latest version of ESLint, but if I'm not mistaken, you're saying it is being reported as an error. It's also confusing that the second example has different configuration than the first - I'm assuming they're the same.

Could you please fill out the issue template that's requested in the comment above mine? This will help us triage and help you better. Thanks!

**Comment by boggddan** [comment]
EsLint v4.1.1
Node v8.0.0
Npm - v5.0.0

The second example works as a corrnet, it does not have an error. The comment remained from the first example.

Question on the first example that there was no mistake EsLint. 
```javascript
const animal = new function() {
  this.name = "Samples";
};
```

Thanks

**Comment by kaicataldo** [comment]
I'm sorry, I'm still not understanding entirely. Would you mind showing code that doesn't error and then code that does with the actual output from running ESLint on the command line?

**Comment by boggddan** [comment]
File with error:
```javascript
/* eslint no-invalid-this: "error", no-unused-vars: 'off'*/
/* eslint-env es6*/
'use strict';

function test() {
  const aObj = new function( text ) {
    this.test = text; // this eslint error "no-ivalid-this"
  }( 'test' );
  return aObj;
}
```

output from command line:
```bash
  7:5  error  Unexpected 'this'  no-invalid-this

✖ 1 problem (1 error, 0 warnings)
```
![2017-07-07_10h28_10](https://user-images.githubusercontent.com/24915025/27947649-6ceaa2ba-62ff-11e7-86ca-d42a2e5d5ead.png)

.eslintrc.js
```javascript
module.exports = {
  env: {
    browser: true,
    es6: true,
    jquery: true,
    node: false
  },
  extends: 'eslint:recommended',
  parserOptions: {
    ecmaVersion: 2017,
    sourceType: 'module',
    ecmaFeatures: {
      experimentalObjectRestSpread: true
    },
  },
  rules: {
  }
};
```

**Comment by mysticatea** [comment]
I'm sorry for late response.

At least, it's working as intended by [the document](http://eslint.org/docs/rules/no-invalid-this). I think we can enhance the rule for function expressions in the callee of new expressions.

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.
